packages/@aws-cdk/region-info/README.md

```diff
@@ -0,0 +1,80 @@
+# AWS Region-Specific Information Directory
+## Usage
+Some information used in CDK Applications differs from one AWS region to
+another, such as service principals used in IAM policies, S3 static website
+endpoints, ...
+
+### Direct Access
+This library offers a primitive database of such information so that CDK
+constructs can easily access regional information. The `Facts` class provides a
+list of known fact names, which can then be used with the `RegionInfo` to
+retrieve a particular value:
+
+```ts
+import { Facts, RegionInfo } from '@aws-cdk/region-info';
+
+const codeDeployPrincipal = RegionInfo.find('us-east-1', Facts.servicePrincipal('codedeploy.amazonaws.com'));
+// => codedeploy.us-east-1.amazonaws.com
+
+const staticWebsite = RegionInfo.find('ap-northeast-1', Facts.s3StaticWebsiteEndpoint);
+// => s3-website-ap-northeast-1.amazonaws.com
+```
+
+### Token Access
+Often, CDK constructs are not able to access the name of the region for which a
+fact is requested (becase this information will not be available before the
+synthesis is underway). To make usage in such scenarios easier, the library
+provides a `RegionInfoToken` class:
+
+```ts
+import { Facts, RegionInfoToken } from '@aws-cdk/region-info';
+
+const staticWebsite = new RegionInfoToken(Facts.s3StaticWebsiteEndpoint);
+```
+
+Tokens initialized in this way will resolve during the systesis phase, according
+to the region name that was determined as part of the synthesis process.
+
+In certain cases, it can be desirable to provide a default value for such tokens
+(for example, where there is a value that is known to work in most regions):
+
+```ts
+// The SNS service principal is almost always "sns.amazonaws.com", so it's a pretty safe default!
+const snsServicePrincipal = new RegionInfoToken(Facts.servicePrincipal('sns.amazonaws.com'), 'sns.amazonaws.com');
+```
+
+## Supplying new or missing information
+As new regions are released, it might happen that a particular fact you need is
+missing from the library. In such cases, the `RegionInfo.register` method can be
+used to inject facts into the database:
+
+```ts
+RegionInfo.register({
+  region: 'bermuda-triangle-1',
+  name: Facts.servicePrincipal('s3.amazonaws.com'),
```

Comment:
Since facts are a strongly-type set, why not just allow people to define the value through a strongly-typed interface:

```ts
RegionInfo.add('bermuda-triangle-1', {
  servicePrincipals: {
    's3.amazonsaws.com': 's3-website.bermuda-triangle-1.nowhere.com'
  }
});
```

Reply:
Backwards compatibility. If I strongly model this, it makes it very easy to break user code that defines overrides because suddenly they don't provide the new key I just introduced.

Reply:
Obvsiouly all properties in the interface that `register`/`add` accepts must be optional so users don't "have" to provide an override (and there are fallback heuristics in the library in case the value is not specified).